synapse/handlers/appservice.py

```diff
@@ -901,3 +902,56 @@ async def claim_e2e_one_time_keys(
                 missing.extend(result[1])
 
         return claimed_keys, missing
+
+    async def query_keys(
+        self, query: Mapping[str, Optional[List[str]]]
+    ) -> Dict[str, Dict[str, Dict[str, JsonDict]]]:
+        """Query application services for device keys.
+
+        Args:
+            query: map from user_id to a list of devices to query
+
+        Returns:
+            A map from user_id -> device_id -> device details
+        """
+        services = self.store.get_app_services()
+
+        # Partition the users by appservice.
+        query_by_appservice: Dict[str, Dict[str, List[str]]] = {}
+        for user_id, device_ids in query.items():
+            if not self.store.get_if_app_services_interested_in_user(user_id):
+                continue
+
+            # Find the associated appservice.
+            for service in services:
+                if service.is_exclusive_user(user_id):
```

Comment:
do the users *have* to be exclusive? This isn't explained here or mentioned in the MSC, so would benefit from an explanation.

Reply:
The [first line of proposal](https://github.com/matrix-org/matrix-spec-proposals/blob/travis/msc/enc-as/key-query/proposals/3984-sending-key-queries-to-appservices.md#proposal) says they must be exclusive:

> the homeserver always proxies the following APIs for an appservice's **exclusive users** to the appservice using the new API described below

Reply:
I've added info to the docstring about this being exclusive users.

Reply:
that was a very subtle mention in the proposal but fair, thanks for noting it in the source

Reply:
Can make it clearer if needed! I think the whole thing only makes sense for exclusive users otherwise you can potentially steal keys?